### ISSUE TITLE
WebGPU rendering backend

### ISSUE THREAD

**Issue by Gerharddc** [slint-ui/slint#171]
This project looks very interesting, but I am wondering if a WebGPU based rendering backend (probably using https://github.com/gfx-rs/wgpu-rs) would not be more appropriate for a modern use case than an OpenGL ES 2.0 one as this would allow for modern standards like Vulkan, Metal and DX12 to be supported automatically.

**Comment by tronical** [comment]
Yes, I think gfx-rs/wgpu-rs is awesome and it'll make sense in the future to use it. At the moment if we switched to gfx-rs we would loose the ability to run our demos in Safari on iOS/iPadOS/macOS because that's limited to WebGL 1. The ability to run our demos in the browser has been of great help, but I believe there's a good chance that upcoming versions of Safari will enable WebGL 2 out of the box.

We're also using the femtovg renderer now, on top of glow. So the effort of switching to gfx-rs will mean contributing to the effort of adding a gfx-rs backend to femtovg (which makes sense, no doubt).

**Comment by Gerharddc** [comment]
Interesting, I have never heard of femtovg (did see nanovg it is based on years ago though), but yeah it obviously makes sense to add it as a back-end there if you are using that for rendering. So if something like femotovg is used then I guess this means that sixtyfps uses a traditional imperative drawing approach instead of a Qt Quick style scene graph approach?

**Comment by tronical** [comment]
The approaches are rather similar:

The qtquick item tree creates a maintains a scene graph node tree at runtime, which is traversed to build GL commands (draw and minimal geometry and texture buffer).

SixtyFPS creates an item tree at compile time that’s equivalent to the scene graph node tree. It’s traversed and femtovg generates list of GL commands (draw, geometry buffers and minimal texture).

So qtquick has slightly better geometry batching (which has advantages but also it’s own caveats) but maintains two trees. SixtyFPS maintains a tree less, does more work ahead of time but femtovg doesn’t do geometry batching.

There might be holes in my understanding, so if I missed something please correct me :)

**Comment by Gerharddc** [comment]
I see. Well yes the main difference that I am thinking about is that in Qt Quick you cannot issue "drawing commands" in order to render elements, you have to manually construct "geometry" and "materials" which more closely maps to GL primitives and I understand that building up a tree of such things is what allows for the efficient batching. 

And I guess the other advantage of defining elements on a more low level like that is that "redrawing" then does not have to be done from scratch. I presume that when using femtovg and you want to for instance change the background colour of a button, then a complete new set of GL commands are calculated and sent to the GPU, whereas in Qt Quick a component could directly inform the rendering engine that it only needs to change one uniform and does not need to replace the entire vertex buffer for instance.

I must admit that I have no idea if what I just described is actually what using femtovg entails, but I would image that it is very hard to make a library that accepts high level drawing commands, and then manages to directly calculate the delta in GL commands between two draw calls with slightly different parameters. Anything I can imagine to implement would simply calculate the GL commands from scratch at each drawing call and if I want to be fancy, I would store the old GL commands and then difference them with the current one in order to at least try and minimize communication with the GPU (I have no idea if anything actually does this).

But then again, forcing component creators to define things on such a low level is a horrible experience for them as opposed to the very pleasant experience of using high level drawing commands.

I know that Flutter manages to achieve very high performance while also using high level drawing calls (in their case directed at Skia), but I have not been able to figure out if there is some serious magic in their system or if repainting from scratch just isn't as bad as it sounds most of the time.

**Comment by tronical** [comment]
I've been thinking a bit more about this and I think we're close to a point where it makes sense to assign a higher priority to this. A few data points:

    * Tickets like #423 will require the GL backend to expose API. Maybe it's better if this were wgpu based.
    * All the wasm platforms we support for demos now also support WebGL 2 (GLES 3.0), which in turn wgpu supports as a backend.

**Comment by Be-ing** [comment]
A use case this would help with is running SixtyFPS applications in a Windows VM from Linux with QEMU-KVM. My SixtyFPS application builds on Windows, but it panics at runtime because it can't create an OpenGL context. I don't have any computers running Windows on bare metal so I'm currently unable to test my application on Windows. There was a [GSOC project](https://gist.github.com/Keenuts/199184f9a6d7a68d9a62cf0011147c0b) in 2017 to write a Windows driver for OpenGL within QEMU VMs, but that has not been completed. However, the wgpu examples run in this environment (albeit slowly):
![image](https://user-images.githubusercontent.com/9455094/152797592-59c8bec2-7535-4415-81af-c7e253261ed0.png)

**Comment by jquesada2016** [comment]
Any updates on this?

**Comment by Horbin-Magician** [comment]
I'm curious, too. Any updates on this?

**Comment by tronical** [comment]
Unfortunately no. Would you have any recommendations for a wgpu based renderer that's on crates.io?

**Comment by tronical** [comment]
I see vello nowadays has updates and looks very promising.

**Comment by Horbin-Magician** [comment]
> I see vello nowadays has updates and looks very promising.

I did some research, and agree that vello is the most appropriate library by far.

**Comment by xb284524239** [comment]
I noticed that the [FemtoVG](https://github.com/femtovg/femtovg/releases/tag/v0.11.0) library has added support for WGPU in version 0.11.0. Does it mean that we can also choose to use the WGPU backend in Slint soon? Because the building speed based on the Skia backend is really, really slow (on my work laptop)....

**Comment by tronical** [comment]
That’s right:). I’ve got a work in progress branch for that for Slint and plan to make a PR. It won’t perform miracles ;-) but it’ll be an opt-in.

**Comment by amitsheokand** [comment]
> That’s right:). I’ve got a work in progress branch for that for Slint and plan to make a PR. It won’t perform miracles ;-) but it’ll be an opt-in.

I wanted to test it out, but It is incomplete, slint::GraphicsAPI still shows NativeOpenGL and WebGL.
we have a [big application](https://aftershoot.com) written in Electron and Rust but we want to port frontend to Slint. 
If It is on roadmap, we can also discuss commercial aspect of it. 
If you can give a timeline that would be great.
Thank you for great work on Slint.

**Comment by tronical** [comment]
I'll rebase my branch on Monday. Do you need a GraphicsAPI / rendering notifier integration? If yes, what kind of functionality do you need?

In theory we could expose wgpu types there, but wgpu apis change, so those parts of the slint api would also not be stable.

Let us know what you need and then we can make a plan :)

**Comment by tronical** [comment]
Ok, we've got `renderer-femtovg-wgpu` as a feature now, to select the WGPU based FemtoVG renderer, and we've got an `unstable-wgpu-24` feature (might become 25, etc.) to expose WGPU based for texture/underlay/overlay integration (see [docs](https://snapshots.slint.dev/master/docs/rust/slint/wgpu_24/)).

I'm going to close this issue, but this shall not be the end of it. I'd love to see a vello based renderer as an option as well, in the future.
